plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/MultiSig.hs

```diff
@@ -0,0 +1,160 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE DeriveGeneric       #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell     #-}
+-- | Implements an n-out-of-m multisig contract.
+module Language.PlutusTx.Coordination.Contracts.MultiSig
+    ( MultiSig(..)
+    , msValidator
+    , msDataScript
+    , msRedeemer
+    , msAddress
+    , lock
+    , initialise
+    , unlockTx
+    -- * Emulator traces
+    , threeOutOfFive
+    ) where
+
+import           Control.Lens
+import           Control.Monad                (void)
+import qualified Data.Map                     as Map
+import qualified Data.Set                     as Set
+import           Data.Foldable                (fold)
+import qualified Ledger.Ada                   as Ada
+import qualified Ledger.Crypto                as Crypto
+import qualified Ledger.Value                 as Value
+import qualified Language.PlutusTx            as P
+import           Ledger                       as Ledger hiding (initialise, to)
+import           Ledger.Validation            as V
+import           Wallet.API                   as WAPI
+import qualified Wallet.Emulator.Types        as EM
+
+data MultiSig = MultiSig 
+                { signatories :: [Ledger.PubKey]
+                -- ^ List of public keys of people who may sign the transaction
+                , requiredSignatures :: Int
+                -- ^ Minimum number of signatures required to unlock
+                --   the output (should not exceed @length signatories@)
+                }
+P.makeLift ''MultiSig
+
+msValidator :: MultiSig -> ValidatorScript
+msValidator sig = 
+  ValidatorScript (Ledger.applyScript mkValidator (Ledger.lifted sig)) where
+    mkValidator = Ledger.fromCompiledCode ($$(P.compile [||
+      let 
+        validate :: MultiSig -> () -> () -> PendingTx -> ()
+        validate (MultiSig keys num) () () p = 
+            let 
+                present = $$(P.length) ($$(P.filter) ($$(V.txSignedBy) p) keys)
+            in
+                if $$(P.geq) present num

```

Comment:
For illustration perhaps it would be nice to log the keys that were present and those that weren't or something?

Reply:
I don't think we can log them from on-chain code (we'd need to to base16-encode them and then convert to a string) but I'll have a look

Reply:
A way to convert bytestrings into strings on chain would be nice.